File: xaiop-sdk/python/src/xaiop/clone.py
"""Deep-clone JSON-compatible values (objects / arrays / scalars)."""

from __future__ import annotations

import copy
import json
from typing import Any


def clone_json(value: Any) -> Any:
    if value is None:
        return value
    t = type(value)
    if t in (str, int, float, bool):
        return value
    if t is not dict and t is not list:
        try:
            return copy.deepcopy(value)
        except Exception:
            return value
    if isinstance(value, list):
        return [clone_json(v) for v in value]
    if isinstance(value, dict) and type(value) is dict:
        return {k: clone_json(v) for k, v in value.items()}
    try:
        return copy.deepcopy(value)
    except Exception:
        try:
            return json.loads(json.dumps(value))
        except Exception:
            return value

File: xaiop-sdk/go/xaiop/clone.go
package xaiop

// CloneJSON deep-clones JSON-compatible values (maps / slices / scalars).
func CloneJSON(value any) any {
	if value == nil {
		return nil
	}
	switch v := value.(type) {
	case string, bool:
		return v
	case int:
		return v
	case int64:
		return v
	case float64:
		return v
	case []any:
		out := make([]any, len(v))
		for i, el := range v {
			out[i] = CloneJSON(el)
		}
		return out
	case map[string]any:
		out := make(map[string]any, len(v))
		for k, el := range v {
			out[k] = CloneJSON(el)
		}
		return out
	default:
		return v
	}
}
